Dont precompile already seen keys, limit epilogue choices

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #124031
* #122825
* #123229
* __->__ #122642
* #124030

Two changes: 
- in epilogue benchmark fusion, only take top 6 choices. There were basically no choices taken after this in HF.
- Share a single precompilation function among matmuls with same key. 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

eellison committed:
Dont precompile already seen keys, limit epilogue choices

[ghstack-poisoned]

eellison committed:
Update on "Dont precompile already seen keys, limit epilogue choices"




[ghstack-poisoned]

eellison committed:
Update on "Dont precompile already seen keys, limit epilogue choices"




[ghstack-poisoned]

shunting314 reviewed (commented):
Can we share a bit of the impact on compilation time and perf? If a dashboard run takes too long, some local run results for a few models (or microbenchmarks) should also be good.

shunting314 commented on torch/_inductor/select_algorithm.py:
```diff
@@ -844,6 +853,8 @@ def __call__(
         def make_benchmark_fn():
             return self.make_benchmark_fn(choices, input_nodes, layout, input_gen_fns)
 
+        inputs_key = repr([self.key_of(x) for x in input_nodes])
```
Give it another thoughts, mm/addmm/bmm happens to have different input metadatas... But would be nice to add op name to the precompile cache key.

eellison replied:
added test, and name and precision in key

eellison committed:
Update on "Dont precompile already seen keys, limit epilogue choices"


Two changes: 
- in epilogue benchmark fusion, only take top 6 choices. There were basically no choices taken after this in HF.
- Share a single precompilation function among matmuls with same key. 

[ghstack-poisoned]

eellison committed:
Update on "Dont precompile already seen keys, limit epilogue choices"


Two changes: 
- in epilogue benchmark fusion, only take top 6 choices. There were basically no choices taken after this in HF.
- Share a single precompilation function among matmuls with same key. 

[ghstack-poisoned]

eellison committed:
Update on "Dont precompile already seen keys, limit epilogue choices"


Two changes: 
- in epilogue benchmark fusion, only take top 6 choices. There were basically no choices taken after this in HF.
- Share a single precompilation function among matmuls with same key. 

[ghstack-poisoned]

eellison committed:
Update on "Dont precompile already seen keys, limit epilogue choices"


Two changes: 
- in epilogue benchmark fusion, only take top 6 choices. There were basically no choices taken after this in HF.
- Share a single precompilation function among matmuls with same key. 

[ghstack-poisoned]

eellison commented:
@pytorchbot merge

eellison committed:
Update on "Dont precompile already seen keys, limit epilogue choices"


Two changes: 
- in epilogue benchmark fusion, only take top 6 choices. There were basically no choices taken after this in HF.
- Share a single precompilation function among matmuls with same key. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison committed:
Update on "Dont precompile already seen keys, limit epilogue choices"


Two changes: 
- in epilogue benchmark fusion, only take top 6 choices. There were basically no choices taken after this in HF.
- Share a single precompilation function among matmuls with same key. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison committed:
Update on "Dont precompile already seen keys, limit epilogue choices"


Two changes: 
- in epilogue benchmark fusion, only take top 6 choices. There were basically no choices taken after this in HF.
- Share a single precompilation function among matmuls with same key. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]
